[EC-271] Refactor CryptoService - move symmetric encryption to EncryptService

eliykat opened the pull request:
## Type of change

<!-- (mark with an `X`) -->

```
- [ ] Bug fix
- [x] New feature development
- [x] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

<!--Describe what the purpose of this PR is. For example: what bug you're fixing or what new feature you're adding-->
#2892 created `EncryptService`, which is intended to house all encryption/decryption methods that are currently contained in `CryptoService`. Importantly, this is intended to be stateless, so it can be used in other contexts where `stateService` is not available.

This PR continues this work by moving all other symmetric encryption/decryption methods to `EncryptService`.

It also adds unit tests, as CryptoService does not have tests at the moment.

This enables #3043, which alleviates performance issues in #1597.

## Code changes

<!--Explain the changes you've made to each file or major component. This should help the reviewer understand your changes-->
<!--Also refer to any related changes or PRs in other repositories-->

Production code:

- **libs/common/src/services/crypto.service.ts**
  - copy the following public methods to encryptService: `encryptToBytes`, `decryptToBytes` and `decryptFromBytes`.
  - copy the private `aesDecryptToBytes` method to encryptService, except for the key resolution logic.
  - delete the private `aesEncrypt` method, it had already been copied to encryptService and was no longer used here after these refactors.
  - move the key resolution logic from the private `aes` methods to the public methods on cryptoService. Then have them call encryptService to do the real work. This effectively leaves cryptoService's functionality unchanged so that existing code continues to work, while extracting the core logic to encryptService.

- **libs/common/src/services/encrypt.service.ts**
  - move the above cryptoService methods to this file.
  - additional refactor: `decryptFromBytes` is fundamentally the same method as `decryptToBytes`, except that it takes an ArrayBuffer instead of an EncString. The EncString object has an interface to get its different parts (e.g. iv, mac, ciphertext), but this logic for the ArrayBuffer was not encapsulated in the same way. Move the code that parses the ArrayBuffer to the `EncArrayBuffer` class. It has the same interface as EncString, now we can combine these methods into a single method: `decryptToBytes`, which takes `EncString | EncArrayBuffer`.

- **libs/common/src/models/domain/encString.ts** - also add some helpers here to give it the same interface as EncArrayBuffer.

Testing code:
- **package.json**
  - install `jest-mock-extended`. This makes it infinitely easier to set up jest mocks, to me it seems like the missing link when using Jest with TS.
  - I was getting regular cache issues when running jest, so added the `--clearCache` argument to the `test:watch` command.
- **libs/common/spec/utils.ts** - add an optional arg to `makeStaticByteArray` so that you can offset its values. Otherwise, the same call will produce arrays with equal values, which could lead to false matches in unit tests.
- **libs/common/spec/matchers/toEqualBuffer.ts** - add a custom Jest matcher to deal with comparing ArrayBuffers. This is used extensively throughout the tests. See the comment in the file for why it's needed.
- **libs/common/spec/test.setup.ts** - bootstrap custom matcher per the docs
- other tests are self-explanatory. Note that the cryptoService test suite is really just a placeholder to encourage more tests in the future (I'll need it shortly anyway).


## Screenshots

<!--Required for any UI changes. Delete if not applicable-->

## Before you submit

<!-- (mark with an `X`) -->

```
- [x] I have checked for **linting** errors (`npm run lint`) (required)
- [x] I have added **unit tests** where it makes sense to do so (encouraged but not required)
- [ ] This change requires a **documentation update** (notify the documentation team)
- [ ] This change has particular **deployment requirements** (notify the DevOps team)
```

coroiu commented on libs/common/spec/services/encrypt.service.spec.ts:
```diff
@@ -0,0 +1,128 @@
+  });
+
+  describe("encryptToBytes", () => {
+    it("encrypts data with provided key", async () => {
```
question: there is only one test-case here, but can this fail in some way?

eliykat replied:
The good thing about this method is that it's fairly flat - there aren't a lot of conditionals (which would require lots of tests).

I have added additional tests for:
* encrypting without a `macKey` - which is one of the few conditionals.
* throwing if no key is provided

coroiu replied:
Looks good!

coroiu commented on libs/common/spec/services/encrypt.service.spec.ts:
```diff
@@ -0,0 +1,128 @@
+  });
+
+  describe("decryptToBytes", () => {
+    const encType = EncryptionType.AesCbc256_HmacSha256_B64;
+
+    const encryptedBytes = makeStaticByteArray(60);
+    const tempArray = new Uint8Array(1 + encryptedBytes.buffer.byteLength);
+    tempArray.set([encType]);
+    tempArray.set(encryptedBytes, 1);
+    const encBuffer = new EncArrayBuffer(tempArray.buffer);
+
+    const key = new SymmetricCryptoKey(makeStaticByteArray(64, 100), encType);
+
+    const decryptedBytes = makeStaticByteArray(10, 200).buffer;
+    const computedMac = new Uint8Array(1).buffer;
```
suggestion: could you describe this setup in some way? Maybe so it reads something like: `given [something something] it decrypts data when provided the correct key`

eliykat replied:
Most of this is to set up the required args for `decryptToBytes`. If I included it in the test names, every test would say `given a key and encrypted data....`  - which goes without saying, they're required.

However, I have refactored this:
* `decryptedBytes` was only used in 1 test, so I've pushed it down
* we can create the `encBuffer` in 1 line. This is because `encType` is the first value, so we can call `makeStaticByteArray` starting at `encType`.

So this should be much clearer now anyway.

coroiu commented on libs/common/spec/services/encrypt.service.spec.ts:
```diff
@@ -0,0 +1,128 @@
+      cryptoFunctionService.hmac.mockResolvedValue(computedMac);
+    });
+
+    it("decrypts data with provided key", async () => {
```
question: Does this refer to the key that was used to encrypt the data? And if yes, what happens when it's the wrong key?

eliykat replied:
Yes, that's right. If it's the wrong key, then the mac check would fails and we return `null` (which is separately tested).

coroiu commented on libs/common/spec/services/encrypt.service.spec.ts:
```diff
@@ -0,0 +1,128 @@
+      expect(actual).toEqualBuffer(decryptedBytes);
+    });
+
+    it("compares macs", async () => {
```
question: Doesn't `returns null if macs don't match` also test the same thing but in a simpler way?

eliykat replied:
`compares macs` asserts that the correct arguments are passed to `cryptoFunctionService.hmac` and `cryptoFunctionService.compare` methods. (which are the lower-level crypto functions that handle this stuff)

`returns null if macs don't match` doesn't check the `cryptoFunctionService` calls, it just mocks its result and makes sure a negative result is properly handled (i.e. an early return of `null`).

I wondered whether these should be combined, but I think it keeps the tests smaller & simpler to have them separate.

So they each test different things, and I think both are required for proper coverage. I'll update the name of `compares macs` to be more descriptive though.

coroiu replied:
Yeah I understand what you're saying and the name `compares macs using CryptoFunctionService` makes a lot more sense!

The reason I brought this up is that I usually try to avoid assert function calls, so whenever I see them I try to see if there are other ways to test. The reason is that I strive to write behavior driven tests and avoid testing implementation details. 

My reasoning is that, if we take `decryptToBytes` as an example: We want to test that mismatching `mac` leads to `null` return. If someone writes an alternative method to `cryptoFunctionService.compare` (maybe a more efficient one, or maybe they install a library) and they modify `decryptToBytes` to use the new function, they could run the tests and see that everything still behaves as expected. However, if tests break they will loose confidence in their refactor. Maybe they think that it's important that the comparison is done specifically using `cryptoFunctionService.compare`, maybe it does something in an unexpected way.

But I can agree that there are some cases where function calls might be part of the actual behavior that we want to test. For example `returns null if macs don't match` checks `expect(cryptoFunctionService.aesDecrypt).not.toHaveBeenCalled();` which I interpret as a test that checks `does not attempt decryption when macs do not match`.

eliykat replied:
PS. your understanding of `return null if macs don't match` is correct. The idea was to make sure the call was returning `null` for the right reason. (it's also possible to call a non-mocked function on a mock class and get `null`!)

coroiu replied:
> We want to make sure it compares macs, in both happy and unhappy code paths. It's not a return value. [...] anyone changing this boundary should be very careful and aware of what they're doing.

After this I think we are in perfect agreement 😄 Thank you for a nice and enlightening conversation! Great job on the PR and hopefully someone more knowledgeable will be able to review the other parts soon!

coroiu commented on libs/common/spec/domain/encArrayBuffer.spec.ts:
```diff
@@ -0,0 +1,59 @@
+    });
+  });
+
+  it("doesn't parse the buffer if invalid", () => {
```
suggestion: Clarify what makes the buffer invalid, eg. `doesn't parse the buffer when buffer is of invalid length`

eliykat replied:
I've:
* updated the test name
* tested length on all 3 encryption types instead of just 1
* added a test for invalid encryption type

Those are all the failures I can think of.

coroiu replied:
Awesome, thanks! The comments are also an appreciated addition!

coroiu commented on libs/common/src/models/domain/encString.ts:
```diff
@@ -119,4 +119,16 @@ export class EncString {
     }
     return this.decryptedValue;
   }
+
+  get ivBytes(): ArrayBuffer {
+    return this.iv == null ? null : Utils.fromB64ToArray(this.iv).buffer;
+  }
+
+  get macBytes(): ArrayBuffer {
+    return this.mac == null ? null : Utils.fromB64ToArray(this.mac).buffer;
+  }
+
+  get ctBytes(): ArrayBuffer {
+    return this.data == null ? null : Utils.fromB64ToArray(this.data).buffer;
+  }
```
question: no tests for these?

eliykat replied:
They seemed like very simple calls to Utils, which is separately tested. I tend to think twice about testing single line methods that just redirect to another class, it feels like there's not much here to test.

This isn't a strongly held position though, what do you think? The tests would be simple enough to add if you think it's worth doing.

coroiu replied:
I was mostly just double checking since you had added so many other tests. If the `Utils` are tested then I totally agree with you, the ROI for adding tests to these functions is not worth it. And even `Utils` weren't tested then the effort would be better spent adding test to them instead of testing these functions!

TL;DR Keep as is :) Not worth the cost of adding and maintaining

coroiu commented on libs/common/src/services/encrypt.service.ts:
```diff
@@ -70,6 +98,29 @@ export class EncryptService implements AbstractEncryptService {
     return this.cryptoFunctionService.aesDecryptFast(fastParams);
   }
 
+  async decryptToBytes(
+    encStringOrBuffer: EncString | EncArrayBuffer,
+    key: SymmetricCryptoKey
+  ): Promise<ArrayBuffer> {
+    if (key == null) {
+      throw new Error("No encryption key provided.");
+    }
```
suggestion: `CryptoService.decryptFromBytes` seems to check if the buffer is `null`, but this method does not, maybe add here so that we don't loose that check when we remove/stop using `CryptoService` for decryption

eliykat replied:
Good pick up, done

coroiu commented on libs/common/src/services/encrypt.service.ts:
```diff
@@ -37,8 +39,34 @@ export class EncryptService implements AbstractEncryptService {
     return new EncString(encObj.key.encType, data, iv, mac);
   }
 
+  async encryptToBytes(plainValue: ArrayBuffer, key: SymmetricCryptoKey): Promise<EncArrayBuffer> {
```
suggestion: add null check to `plainValue`?

eliykat replied:
This wasn't in the code originally, so I checked: it is possible to encrypt a `null` value :D I don't know if we do this, but it's at least possible, so I don't want to add it.

coroiu reviewed (commented):
Very appreciated with a smaller PR instead of the `huge scary PR` that you mentioned 😄 

I'm still pretty new and haven't had the chance to work with the crypto parts of the application yet, so I'm not really able to review the actual crypto logic. But I added some comments and questions, mainly surrounding the tests.

eliykat commented:
I appreciate the review @coroiu! Just let me know once I've addressed your feedback, but don't properly "approve" it given that it needs someone else to look at it as well.

Hinton commented on package.json:
```diff
@@ -18,7 +18,7 @@
     "prettier": "prettier --write .",
     "test": "jest",
-    "test:watch": "jest --watch",
+    "test:watch": "jest --clearCache && jest --watch",
```
Do we want to clear the cache here? I would typically not expect it to be cleared.

eliykat replied:
It's for convenience, I was finding that Jest would fall out of sync and start running old code after a while. (i.e. it wouldn't be running my latest saved changes.) This let me re-run the command to fix it.

I assume that if you're running `test:watch` then you're probably leaving it running and not restarting it frequently (making use of the `watch` functionality). So a slight delay on start to clear the cache seems okay.

I'm not strongly wedded to it, it just helped me.

Hinton commented on libs/common/src/services/encrypt.service.ts:
```diff
@@ -86,6 +141,44 @@ export class EncryptService implements AbstractEncryptService {
   }
 
+  private async aesDecryptToBytes(
+    encType: EncryptionType,
+    data: ArrayBuffer,
+    iv: ArrayBuffer,
+    mac: ArrayBuffer,
```
Thoughts on having this only accept an `EncString` or `EncArrayBuffer` (Or preferably make an interface of those two classes)? Those two should already know how to extract the `data`, `iv` and `mac`.

eliykat replied:
This private method is only called from `decryptToBytes`, which is almost a bare call to the private method. So I've done you one better and just combined the private method into the public method (which already takes the whole object).

I've kept the public interface as accepting `EncString | EncArrayBuffer`, TS union types seem convenient here to avoid creating/extending new interfaces. I don't think it needs to be generalized beyond that union.

I'll also note that `EncryptedObject` should be refactored into an `EncArrayBuffer`, it's not really needed and the name is confusing things. But that part isn't fully unit tested (not part of this PR), so I've left it for now.

Hinton commented:
@MGibson1 mind reviewing this more in-depth? I haven't looked at the crypto service in a while.

eliykat commented:
> We should create a tech debt item for the refactors to the crypto service, and link to it in the new comments.

I'm not sure whether these comments assume too much, maybe these `cryptoService` methods won't go away? Interested to hear @MGibson1's view.

eliykat commented:
Applying the `hold` label so that I don't merge until after `rc` is cut.  But reviews in the meantime are welcome.

MGibson1 commented on libs/common/src/models/domain/encArrayBuffer.ts:
```diff
@@ -1,3 +1,53 @@
+        this.ctBytes = encBytes.slice(17).buffer;
+        break;
+      default:
+        return;
```
It is preferrable to have types throw that cannot initialize rather than silently not populate.

The silent model produces null errors down the line that are difficult to trace. Loud errors tell you exactly where the problem is happening.

eliykat replied:
Good point about not propagating nulls. I was focusing on replicating current behaviour, but this shouldn't cause any issues, it should just enforce current assumptions.

What do you think about parsing in a private method instead, which is called by getters? e.g.

```ts
private parse(): { iv: ArrayBuffer, mac: ArrayBuffer, data: ArrayBuffer } {
  // current parsing logic here, returns the object described above
  // this keeps the parsing logic all in one place, and array slicing is cheap
}

get macBytes() {
  return this.parse().mac;
}

get ivBytes() {
 return this.parse().iv;
}
```

We could still check for a valid input in the constructor and throw if required.

I'm also not 100% sure that the current logic won't create duplicate arrays in memory. Depends on Javascript memory management. But I think this would be a bit safer in that respect as well.

MGibson1 replied:
We spoke on the phone about this, but the call to make is this: Should this class exist if it isn't parseable (and therefore not decryptable)? There's an argument either way. I'm OK with either.

eliykat replied:
I'm glad you mentioned this, because it made me realise that I've broken existing behaviour.

Currently, if the `ArrayBuffer` can't be parsed, `cryptoService.decryptFromBytes` returns `null`, and the caller checks for this.

Now that it's moved to the constructor, the ctor will always return a not-null value, so these null checks will no longer work.

I've changed this to throw explicitly. `cryptoService.encryptToBytes` wraps it in a `try/catch` and returns `null` to preserve existing behaviour.

I also prefer the explicit throw generally, we overuse `null` for this at the moment generally.

eliykat replied:
I also added a test for `CryptoService` to make sure this works, and as a form of documenting this behaviour.

MGibson1 commented on libs/common/src/services/encrypt.service.ts:
```diff
@@ -70,6 +98,55 @@ export class EncryptService implements AbstractEncryptService {
+    }
+
+    if (key.macKey != null && encThing.macBytes != null) {
+      const macData = new Uint8Array(encThing.ivBytes.byteLength + encThing.ctBytes.byteLength);
+      macData.set(new Uint8Array(encThing.ivBytes), 0);
+      macData.set(new Uint8Array(encThing.ctBytes), encThing.ivBytes.byteLength);
```
It looks like you're decrypting these two 'Things' (great name 😆 ) based off an assumed shared interface. Let's make that explicit. `DecryptableInterface`, perhaps?

eliykat replied:
I went with `IEncrypted`, I think that's the interface naming convention we discussed in PIQ, and it won't get confused with objects that have a `decrypt` method.

MGibson1 commented on libs/common/src/services/encrypt.service.ts:
```diff
@@ -70,6 +98,55 @@ export class EncryptService implements AbstractEncryptService {
+      return null;
+    }
+
+    if (key.macKey != null && encThing.macBytes != null) {
+      const macData = new Uint8Array(encThing.ivBytes.byteLength + encThing.ctBytes.byteLength);
+      macData.set(new Uint8Array(encThing.ivBytes), 0);
+      macData.set(new Uint8Array(encThing.ctBytes), encThing.ivBytes.byteLength);
+      const computedMac = await this.cryptoFunctionService.hmac(
+        macData.buffer,
+        key.macKey,
+        "sha256"
+      );
+      if (computedMac === null) {
+        return null;
+      }
+
+      const macsMatch = await this.cryptoFunctionService.compare(encThing.macBytes, computedMac);
+      if (!macsMatch) {
+        this.logMacFailed("mac failed.");
+        return null;
+      }
+    }
```
extract to `macsMatch` method.

eliykat replied:
There are 2 mac comparisons in `EncryptService`, but they are slightly different: one uses `cryptoService.compare` and the other uses `cryptoService.compareFast`. So they can't be combined into a single private method.

MGibson1 reviewed (changes requested):
Just a few suggestion, nothing big.

This is really nice!

eliykat commented on libs/common/src/models/domain/encArrayBuffer.ts:
```diff
@@ -1,3 +1,53 @@
-  constructor(public buffer: ArrayBuffer) {}
+  readonly encryptionType: EncryptionType = null;
+
+  readonly ctBytes: ArrayBuffer = null;
```
I went with `dataBytes` to align with existing naming convention of `EncString`. And it's a bit more intuitive than `ct`.

eliykat commented on libs/common/src/models/domain/encArrayBuffer.ts:
```diff
@@ -1,3 +1,53 @@
+      case EncryptionType.AesCbc128_HmacSha256_B64:
+      case EncryptionType.AesCbc256_HmacSha256_B64:
+        if (encBytes.length <= 49) {
+          // 1 + 16 + 32 + ctLength
```
Good suggestion, I think the result is a little harder to read at a glance, but easier to properly understand.

It also revealed an off-by-one error in my "invalid length" tests!

eliykat commented on libs/common/src/abstractions/crypto.service.ts:
```diff
@@ -80,7 +80,7 @@ export abstract class CryptoService {
   decryptToBytes: (encString: EncString, key?: SymmetricCryptoKey) => Promise<ArrayBuffer>;
   decryptToUtf8: (encString: EncString, key?: SymmetricCryptoKey) => Promise<string>;
-  decryptFromBytes: (encBuf: ArrayBuffer, key: SymmetricCryptoKey) => Promise<ArrayBuffer>;
+  decryptFromBytes: (buffer: ArrayBuffer, key: SymmetricCryptoKey) => Promise<ArrayBuffer>;
```
I want to maintain the legacy interface (note that this is on `CryptoService`). It receives an `ArrayBuffer` and constructs an `EncArrayBuffer`, then passes it to `EncryptService` - providing a bridge for current calls.

The new interface on `EncryptService` does require an `EncArrayBuffer`.

MGibson1 replied:
One of the major reasons for the monorepo was to be able to change these interfaces easily. I don't see this as a good reason for keeping these methods around or static.

However, this particular method is also handling the legacy key thing, which is also largely unsolved in this PR right now.

I'd like to see the method updated, but I'm not going to insist on it.

eliykat replied:
I was treating it like the other interfaces here - we're generally not changing existing callers. But, this is only called from a few places, which were easy enough to update, so it seemed faster to just do it ;) See the latest commit.

eliykat commented on libs/common/spec/matchers/toEqualBuffer.ts:
```diff
@@ -0,0 +1,34 @@
+/**
+ * The inbuilt toEqual() matcher will always return TRUE when provided with 2 ArrayBuffers.
+ * This is because an ArrayBuffer must be wrapped in a new Uint8Array to be accessible.
+ * This custom matcher will automatically instantiate a new Uint8Array on the recieved value
+ * (and optionally, the expected value) and then call toEqual() on the resulting Uint8Arrays.
+ */
```
Glad you like it :D

eliykat commented on libs/common/spec/utils.ts:
```diff
@@ -28,10 +28,11 @@ export function mockEnc(s: string): EncString {
+export function makeStaticByteArray(length: number, start = 0) {
   const arr = new Uint8Array(length);
+  let count = start;
   for (let i = 0; i < length; i++) {
-    arr[i] = i;
+    arr[i] = count++;
```
> ```ts
> for (let i = start; i < start + length ; i++) {
>   arr[i] = i;
> }
> ```

If you call this with "length of 10, starting at 20" then it's going to start inserting values at index 20, right?

> or yet more cryptically, since the keys of an Array are just integers:
> 
> ```ts
> Array.from(Array(length).keys()).map(i => start + i)
> ```

This is nice 😁 but, as you say, cryptic. I think it will just confuse people.

eliykat replied:
But we can do

```ts
  for (let i = 0; i < length; i++) {
    arr[i] = start + i;
  }
```

Sometimes it's the simple things 😁

MGibson1 replied:
> If you call this with "length of 10, starting at 20" then it's going to start inserting values at index 20, right?

lol yeah. I realized that at the time and for some reason thought we wanted it. oh well, thanks for questioning it

eliykat commented on libs/common/src/services/crypto.service.ts:
```diff
@@ -513,33 +512,19 @@ export class CryptoService implements CryptoServiceAbstraction {
-   */
+  // ---ENCRYPTION---
+  // These methods resolve the key and then call encryptService.
+  // This should be refactored so that callers get their own key and give it to encryptService directly,
+  // then these methods should be removed.
```
I've removed this comment altogether. As I think you previously pointed out, we can't deprecate these methods until we provide a way for callers to get their own encryption key, _and_ deal with legacy ETM keys. (We basically need to make `getKeyForEncryption` public and then move `resolveLegacyKey` down to `EncryptService`, to provide an equivalent alternative code path. Which I have ready in a local branch... >_>)

I wanted to provide context here, but I think it's just confusing. I suggest we just hold off on deprecating etc. until we have a workaround.

eliykat commented on libs/common/src/services/crypto.service.ts:
```diff
@@ -513,33 +512,19 @@ export class CryptoService implements CryptoServiceAbstraction {
-   * All callers should use this service to grab the relevant key and use encryptService for encryption instead.
-   * This method will be removed once all existing code has been refactored to use encryptService.
-   */
+  // ---ENCRYPTION---
+  // These methods resolve the key and then call encryptService.
+  // This should be refactored so that callers get their own key and give it to encryptService directly,
+  // then these methods should be removed.
```
Same as above (https://github.com/bitwarden/clients/pull/3042#discussion_r927316620)

eliykat commented on libs/common/src/services/crypto.service.ts:
```diff
@@ -554,6 +539,11 @@ export class CryptoService implements CryptoServiceAbstraction {
     return new EncString(EncryptionType.Rsa2048_OaepSha1_B64, Utils.fromBufferToB64(encBytes));
   }
 
+  // ---DECRYPTION---
+  // These methods resolve the key and then call encryptService.
+  // This should be refactored so that callers get their own key and give it to encryptService directly,
+  // then these methods should be removed.
+
```
Same as above (https://github.com/bitwarden/clients/pull/3042#discussion_r927316620)

eliykat commented on libs/common/src/services/crypto.service.ts:
```diff
@@ -554,6 +539,11 @@ export class CryptoService implements CryptoServiceAbstraction {
   }
 
+  // ---DECRYPTION---
+  // These methods resolve the key and then call encryptService.
+  // This should be refactored so that callers get their own key and give it to encryptService directly,
```
Same as above (https://github.com/bitwarden/clients/pull/3042#discussion_r927316620)

MGibson1 commented on libs/common/src/models/domain/encArrayBuffer.ts:
```diff
@@ -1,9 +1,18 @@
+import { IEncrypted } from "@bitwarden/common/interfaces/IEncrypted";
 
-export class EncArrayBuffer {
+const encTypeLength = 1;
+const ivLength = 16;
+const macLength = 32;
+const minDataLength = 1;
```
Please use CONSTANT_CASE for global constants: [eslint naming rule](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/naming-convention.md#naming-convention)

MGibson1 commented on libs/common/src/models/domain/encArrayBuffer.ts:
```diff
@@ -1,9 +1,18 @@
+const macLength = 32;
+const minDataLength = 1;
+
+const decryptionError =
```
I don't see why this is a constant outside of the class.

eliykat replied:
I think of it as a constant, same as the other constants here. But I have no strong feelings about it, I've refactored it into a private method instead as an alternative way to organise it - `throwDecryptionError`.

MGibson1 replied:
I'm fine with it as a `const string`, it's just that it should belong to the only class that throws it.

A guard method is also fine

MGibson1 commented on libs/common/src/services/crypto.service.ts:
```diff
@@ -614,7 +604,12 @@ export class CryptoService implements CryptoServiceAbstraction {
+    }
 
     key = await this.getKeyForEncryption(key);
     key = await this.resolveLegacyKey(encBuffer.encryptionType, key);
```
I thought you'd said this was going away in favor of slicing the current key.

MGibson1 replied:
If we get rid of this, the whole of cryptoService's decrypt stuff can go away

MGibson1 reviewed (changes requested):
One variable name convention issue that needs to be updated.

Other things I'd like to see:
* Resolve our little disagreement on cryptoService interface.
* I would like to get the new method of legacy key retrieval integrated into this PR.

eliykat commented:
> I would like to get the new method of legacy key retrieval integrated into this PR.

Because this is fairly involved logic (and nobody likes touching it), I intentionally kept this PR small and have been taking this in small chunks. All these changes are (almost) approved, so I'd much prefer to create a separate PR for additional (if related) changes. And that gets this refactor back into `master` for visibility with the rest of the team.

I've created a separate PR here for the key getter + legacy key changes: https://github.com/bitwarden/clients/pull/3170. But let me know if you see any problem with this plan.

eliykat commented on libs/common/src/models/domain/encArrayBuffer.ts:
```diff
@@ -1,3 +1,73 @@
+  }
+
+  static async fromResponse(response: {
+    arrayBuffer: () => Promise<ArrayBuffer>;
```
This interface looks a little weird, but node has its own Response type definition  which we can't access here in common. So I've described it this way as well, which works for the TS and Node response types.

MGibson1 replied:
I like it

MGibson1 commented on apps/cli/src/services/nodeEnvSecureStorage.service.ts:
```diff
@@ -63,10 +64,8 @@ export class NodeEnvSecureStorageService implements AbstractStorageService {
-        Utils.fromB64ToArray(encValue).buffer,
-        sessionKey
-      );
+      const encBuf = EncArrayBuffer.fromB64(encValue);
```
Might need a try/catch around this since we decided to make the constructor throw.

eliykat replied:
This is already within a try/catch that returns a decryption error, which seemed fine.

MGibson1 replied:
yup my mistake. I just didn't see it since it's a relatively large try block. thanks

MGibson1 reviewed (changes requested):
One more concern about a possibly new throw, otherwise I think this is ready to go.
